### DISCUSSION TITLE
Unable to catch `Exception` via exception_handlers since FastAPI 0.70.0

### DISCUSSION THREAD

**Discussion by manlix** [fastapi/fastapi#8647] [Questions] [answered]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
################################################
#
# Pay attention: below is Python 3.10 syntax
#
################################################

from fastapi import FastAPI, Request
from fastapi.responses import JSONResponse
from pydantic import ValidationError

async def internal_server_error(request: Request, exc: Exception | ValidationError) -> "JSONResponse":
    return JSONResponse(content={"detail": "Internal Server Error"}, status_code=500)

app = FastAPI(
    exception_handlers={
        Exception: internal_server_error,
        ValidationError: internal_server_error,
    },
)

@app.get("/call_me")
async def raise_exception() -> None:
    raise Exception("I am exception.")
```

### Description

* Open the browser and call the endpoint: `/call_me`
* It returns **plain text**: `Internal Server Error`
* But I excepted it return **JSON**: `{"detail": "Internal Server Error"}`

### Operating System

Linux

### FastAPI Version

0.70.0

### Python Version

Python 3.10.0

### Additional Context

**It works fine on FastAPI 0.68.2.**
Issue is also affected 0.69.0

**Comment by Kludex** [comment]
The Exception cannot be used for this purpose. You can inherit it tho

**Comment by manlix** [comment]
> The Exception cannot be used for this purpose. You can inherit it tho

It works fine in 0.68.2

My use case: I`m trying to catch **all unknown exceptions** in code and responsing with JSON =)

I think issue is relating to [0.69.0](https://github.com/tiangolo/fastapi/releases/tag/0.69.0) (with new Starlette 0.15.0 and AnyIO)

**Comment by Kludex** [comment]
You're right. I guess I put in my mind that is a "cannot" because Starlette will log the traceback if you use try to handle `Exception`. 

In any case, I cannot reproduce on Python 3.8, jfyk.

**Comment by ElementalWarrior** [comment]
As a side note, I can't seem to catch the exceptions from a websocket endpoint in fastapi==0.68.2

```
@router.websocket("/ws")
async def websocket_endpoint(websocket: WebSocket, config: BaseConfig = Depends(get_config)):
    raise ValueError("something")
```
The stack trace
```
websocket_1  | ERROR:    Exception in ASGI application
websocket_1  | Traceback (most recent call last):
websocket_1  |   File "/venv/lib/python3.8/site-packages/uvicorn/protocols/websockets/websockets_impl.py", line 203, in run_asgi
websocket_1  |     result = await self.app(self.scope, self.asgi_receive, self.asgi_send)
websocket_1  |   File "/venv/lib/python3.8/site-packages/uvicorn/middleware/proxy_headers.py", line 75, in __call__
websocket_1  |     return await self.app(scope, receive, send)
websocket_1  |   File "/venv/lib/python3.8/site-packages/fastapi/applications.py", line 208, in __call__
websocket_1  |     await super().__call__(scope, receive, send)
websocket_1  |   File "/venv/lib/python3.8/site-packages/starlette/applications.py", line 112, in __call__
websocket_1  |     await self.middleware_stack(scope, receive, send)
websocket_1  |   File "/venv/lib/python3.8/site-packages/starlette/middleware/errors.py", line 146, in __call__
websocket_1  |     await self.app(scope, receive, send)
websocket_1  |   File "/venv/lib/python3.8/site-packages/starlette/middleware/cors.py", line 70, in __call__
websocket_1  |     await self.app(scope, receive, send)
websocket_1  |   File "/venv/lib/python3.8/site-packages/starlette/exceptions.py", line 58, in __call__
websocket_1  |     await self.app(scope, receive, send)
websocket_1  |   File "/venv/lib/python3.8/site-packages/starlette/routing.py", line 580, in __call__
websocket_1  |     await route.handle(scope, receive, send)
websocket_1  |   File "/venv/lib/python3.8/site-packages/starlette/routing.py", line 297, in handle
websocket_1  |     await self.app(scope, receive, send)
websocket_1  |   File "/venv/lib/python3.8/site-packages/starlette/routing.py", line 68, in app
websocket_1  |     await func(session)
websocket_1  |   File "/venv/lib/python3.8/site-packages/fastapi/routing.py", line 273, in app
websocket_1  |     await dependant.call(**values)
websocket_1  |websocket_1  | ERROR:uvicorn.error:Exception in ASGI application
websocket_1  | Traceback (most recent call last):
websocket_1  |   File "/venv/lib/python3.8/site-packages/uvicorn/protocols/websockets/websockets_impl.py", line 203, in run_asgi
websocket_1  |     result = await self.app(self.scope, self.asgi_receive, self.asgi_send)
websocket_1  |   File "/venv/lib/python3.8/site-packages/uvicorn/middleware/proxy_headers.py", line 75, in __call__
websocket_1  |     return await self.app(scope, receive, send)
websocket_1  |   File "/venv/lib/python3.8/site-packages/fastapi/applications.py", line 208, in __call__
websocket_1  |     await super().__call__(scope, receive, send)
websocket_1  |   File "/venv/lib/python3.8/site-packages/starlette/applications.py", line 112, in __call__
websocket_1  |     await self.middleware_stack(scope, receive, send)
websocket_1  |   File "/venv/lib/python3.8/site-packages/starlette/middleware/errors.py", line 146, in __call__
websocket_1  |     await self.app(scope, receive, send)
websocket_1  |   File "/venv/lib/python3.8/site-packages/starlette/middleware/cors.py", line 70, in __call__
websocket_1  |     await self.app(scope, receive, send)
websocket_1  |   File "/venv/lib/python3.8/site-packages/starlette/exceptions.py", line 58, in __call__
websocket_1  |     await self.app(scope, receive, send)
websocket_1  |   File "/venv/lib/python3.8/site-packages/starlette/routing.py", line 580, in __call__
websocket_1  |     await route.handle(scope, receive, send)
websocket_1  |   File "/venv/lib/python3.8/site-packages/starlette/routing.py", line 297, in handle
websocket_1  |     await self.app(scope, receive, send)
websocket_1  |   File "/venv/lib/python3.8/site-packages/starlette/routing.py", line 68, in app
websocket_1  |     await func(session)
websocket_1  |   File "/venv/lib/python3.8/site-packages/fastapi/routing.py", line 273, in app
websocket_1  |     await dependant.call(**values)
websocket_1  |   File "./websocket/api/endpoints/websockets.py", line 190, in websocket_endpoint
websocket_1  |     raise ValueError("something")
websocket_1  | ValueError: something   File "./websocket/api/endpoints/websockets.py", line 190, in websocket_endpoint
websocket_1  |     raise ValueError("something")
websocket_1  | ValueError: something
```
App creation
```


    exception_handlers = {
        Exception: unhandled_exception_handler,
        HTTPException: unhandled_exception_handler,
    }

    app = FastAPI(title="websocket server", exception_handlers=exception_handlers)
```

**Comment by lephuongbg** [comment]
Same use case here, wanting to always return JSON even with generic exception. It seems to because of new starlette version https://github.com/encode/starlette/issues/1175.

**Comment by InesIvanova-dev** [comment]
Same here. 
I noticed that when I set up Github actions and I haven't listed fastapi version in the requirements file (which locally was 0.65.0) and on the CI it was 0.70.0). I am using httpx AsyncClient for my tests and I also have tests where I mock exceptions (pure Python exceptions) so my CI started to fail.

**Comment by yo-main** [comment]
Same here. Pinning starlette to the version 0.14.2 fixed the issue.

**Comment by gromsterus** [comment]
Fixed by https://github.com/encode/starlette/pull/1262.

**Comment by alexiri** [comment]
I'm eagerly looking forward to https://github.com/tiangolo/fastapi/pull/4145 getting merged. :smile:

**Comment by Kludex** [comment]
No... This is not fixed...

**Comment by phillipuniverse** [comment]
@Kludex did you validate against the [just-released FastAPI 0.71.0](https://github.com/tiangolo/fastapi/releases/tag/0.71.0) and confirm this is still an issue?

**Comment by ca-simone-chiorazzo** [comment]
@phillipuniverse Using FastAPI 0.73.0 we're still facing such an issue

**Comment by JarroVGIT** [comment]
Using 0.78, I cannot reproduce. I see a JSON response in my browser using the code in OP.

**Comment by ca-simone-chiorazzo** [comment]
We're facing this issue in tests, using the AsyncClient. Could you please try that?

**Comment by Kludex** [comment, marked as answer]
> We're facing this issue in tests, using the AsyncClient. Could you please try that?

Would you mind providing a minimal reproducible example? 🙏

**Comment by amymb** [comment]
Facing a similar issue using FastAPI 0.80 and Starlette 0.19.1. The FastAPI app handles the Exception and returns JSON, but logs a stack trace with the following code:

```
@app.exception_handler(Exception)
async def generic_exception_handler(request: Request, exc: Exception) -> JSONResponse:
  return JSONResponse(content={"detail": "Internal Server Error"}, status_code=500)
```

The stack trace is identical to [the above](https://github.com/tiangolo/fastapi/issues/4025#issuecomment-939044767) (only containing my unhandled error instead of `ValueError`). 

Interestingly, if I convert the exception handler to middleware, there's no logged stack trace. I'd rather use the error handler, though.

```
@app.middleware("http")
async def generic_exception_handler(request: Request, call_next) -> JSONResponse:
  try:
    return await call_next(request)
  except Exception as exc:
    return JSONResponse(content={"detail": "Internal Server Error"}, status_code=500)
```

Problem: an ASGI error gets logged with a stack trace when I try to handle Exception with an exception handler.
What I would like: To handle Exception and only see the logs I put in my exception handler.

**Comment by n00b42** [comment]
Still an issue in 0.109.0

Example
~~~
@app.get("/raise")
async def raise_exception():
    raise Exception("Hello")
@app.exception_handler(Exception):
    return Response(status_code=status.HTTP_500_INTERNAL_SERVER_ERROR)
~~~

The route returns 500 as expected, but the exception is still printed with "Exception in ASGI application"

**Reply by Kludex** [reply]
That's expected.

**Reply by luliangce** [reply]
> That's expected.

so we have to catch unexcepted Exception with middleware, In my case, I have to clear these stack

**Comment by dan5e3s6ares** [comment]
Still an issue in "0.111.0"

**Comment by jonathan-h-grebe** [comment]
Still an issue in "0.112.0"

**Comment by DeoLeung** [comment]
I have an exception handler
```python

@app.exception_handler(RCError)
async def value_error_exception_handler(req: Request, exc: RCError):
  """handle rc error"""
  ...
  if req.scope['type'] == 'websocket':
    # not sure how to handle error here?
    # now if before ws.accept() it returns 403
    # after ws.accept, it returns 500
    raise WebSocketException(status.WS_1006_ABNORMAL_CLOSURE, result)
  else:
    return ORJSONResponse(result, status_code=exc.status_code)
  ```
  
  could distinguish where the exception comes from, but not sure how to handle it
